### PR TITLE
Developer documentation and dev environment setup

### PR DESCRIPTION
**Pull request by lwgordonimos**
This script shortcuts the setup of a virtual environment containing the package dependencies. The idea being that you can just checkout python-aodndata, run the script and have it setup and ready to develop in (e.g. from PyCharm). Could be tided up in future in a few ways, but does the job for now (assuming boto3 and wget are installed).

```bash
$ git clone git@github.com:aodn/python-aodndata.git && cd python-aodndata
$ ./setup_virtualenv.sh 
Downloading dependencies...
Creating virtual environment...
Installing dependencies into virtual environment...
Virtual environment successfully created at: python-aodndata-virtualenv

To use:
  * Configure PyCharm project interpreter as: /tmp/python-aodndata/python-aodndata-virtualenv/bin/python
  * Activate in shell environment: $ source /tmp/python-aodndata/python-aodndata-virtualenv/bin/activate
$ source /tmp/python-aodndata/python-aodndata-virtualenv/bin/activate
(python-aodndata-virtualenv) $ python setup.py test
[snipped]
Ran 79 tests in 8.359s

OK
```

This came out of actually trying to document a simple setup process, so I automated as much as possible...

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -14,6 +14,9 @@ before_install:
   - python get_latest_artifact.py --extension .whl --job compliance_checker_prod
   - python get_latest_artifact.py --extension .whl --job cc_plugin_imos_prod
 
+env:
+  - MPLBACKEND=agg
```
Comment (mhidas) [inline .travis.yml]: Should this be in a separate PR, or is it actually related?

**Comment by mhidas** [comment]
I assume some of the documentation you're writing, or at least a link to it, will be added to this repo's README.md?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+# User Guide
+
+It is highly recommended to use the [PyCharm IDE](https://www.jetbrains.com/pycharm/) for developing on this code base, as it provides many useful features out of the box, such as good unit test integration, real time code quality checking, automatic completion, and the usual basic IDE debugging capabilities such as setting breakpoints and syntax highlighting.
+
+## Setting up the development environment
+
+### Pre-requisites
+1. Ensure the [aodn/chef](https://github.com/aodn/chef) repository is checked out (directory will be referred to as **${CHEF_REPO}**), and the usual pre-requisites for running the PO box are met
+1. Ensure that your GitHub keys are in place so that you have write access to the **python-aodndata** repository on GitHub
+1. Provision the PO box using the **bin/po-box.sh** script, so that the required repositories are checked out and the PO box is ready for use
```
Comment (mhidas) [inline README.md]: Currently we still have to use the **po_pipeline_2** box (launched with `bin/po-pipeline-2-box.sh`). Merging this back into the original **po** box is TBD.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+ ```
+
+### Handler quick start
+
+1. If there isn't already a relevant facility subpackage under **aodndata**, create one (in accordance with the preferred [naming conventions](https://www.python.org/dev/peps/pep-0008/#package-and-module-names)). A package is a directory containing an (often empty) **\_\_init\_\_.py** file, and allows clean arrangement of code into a namespace. The structure is arbitrary, however if in doubt, you may wish to create a **handlers.py** module in the directory to contain the handler code as a starting point:
+
+    ```bash
+    ${CHEF_REPO}/src/python-aodndata/aodndata$ find myfacility/
+    myfacility/
+    myfacility/__init__.py
```
Comment (mhidas) [inline README.md]: Could be worth adding '[empty file]' after __init.py just to be clear

File: README.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+This allows for a much more rapid turnaround in getting your "work in progress" code to actually run in a live development environment (in this case, the PO box).
+
+Note: the following assumes that you are running the PO box from the same **${CHEF_REPO}** as referred to above in the IDE setup
+
+1. Add a watch configuration to the **imos_po_watches** databag in Chef. A watch configuration defines:
+
+    1. **path**: types: JSON=array, Python=list : list of incoming directory paths to be watched and having incoming files routed to this pipeline
+    1. **handler**: types JSON=string, Python=str : name of the handler class to use (the handler class is resolved by looking for this string in the **pipeline.handlers** entry point group and retrieving the corresponding handler object)
+    1. **params**:  types JSON=object, Python=dict : passed directly to the handler class **_\_\_init\_\__** method as **\*\*kwargs** in order to configure the handler parameters
```
Comment (mhidas) [inline README.md]: Maybe add a note that the list of available arguments and their meanings can be seen in the [HandlerBase code](https://github.com/aodn/python-aodncore/blob/master/aodncore/pipeline/handlerbase.py#L148)
It might be worth including a few more of the commonly used ones in the example below (e.g. "notify_params", "include_regexes", "dest_path_function")

Comment (lwgordonimos) [reply README.md]: Definitely needs that documentation, but like you mentioned I was going to put that in aodncore documentation and link to it. Just so that if changed, it doesn't need PRs in multiple repos. 

I think some examples would be good though, so will add them. I can't see include_regexes of dest_path_function ever changing, so should be fine to mention them

File: README.md (hunk 2 above)
Comment (lwgordonimos) [reply README.md]: Yep, I wrote it from the assumption that the merge had occurred.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            handler = self.run_handler_with_exception(ComplianceCheckFailedError, NOT_A_NETCDF)
+            pass
+
+    ```
+
+1. You can now test your handler by simply running the unit tests. There are several ways to run them in the IDE, but you can get fine-grained control over which tests are run by opening the test module, and right-clicking on the test class and clicking _'Run unittests for Unittests for test_handlers.TestMyFacilityHandler'_, or even the individual test methods. This enables the handler class itself to be largely configured
+and tested before leaving the IDE, and proceeding on to integration testing, to run the handler in a "deployed" context.
+
+1. For further documentation relating to available handler parameters, and how a handler class works, refer to the upstream [aodncore documentation](https://github.com/aodn/python-aodncore). 
+Handler parameters consist of a single positional parameter, which is always 'input_file', a 'config' object (set automatically in both unittests and when deployed), an optional Celery task parameter, set when run under a Celery task, and a series of keyword arguments to control
```
Comment (mhidas) [inline README.md]: ... (unfinished sentence). "... a series of keyword arguments to control the behaviour of individual handling steps"?

File: README.md (hunk 4 above)
Comment (mhidas) [reply README.md]: :+1: I wasn't suggesting duplicating the aodncore documentation here, just linking to it.
